File: 2025/12/03/structured-outputs-llms-reliable-json/python/src/schema.py
"""Schema definitions using Pydantic."""
from pydantic import BaseModel, Field
from enum import Enum


class Category(str, Enum):
    """Task category enum."""
    BUG = "bug"
    FEATURE = "feature"
    QUESTION = "question"
    OTHER = "other"


class TaskTriage(BaseModel):
    """Task triage schema for categorizing and prioritizing issues."""
    category: Category = Field(description="Category of the task")
    priority: int = Field(ge=1, le=5, description="Priority from 1 (low) to 5 (critical)")
    needs_human: bool = Field(description="Whether this task requires human intervention")
    summary: str | None = Field(None, description="Brief summary of the issue")
    
    class Config:
        extra = "forbid"  # Reject extra fields


class UserInfo(BaseModel):
    """User information extraction schema."""
    name: str = Field(description="User's full name")
    email: str = Field(description="User's email address")
    age: int | None = Field(None, ge=0, le=150, description="User's age if mentioned")
    
    class Config:
        extra = "forbid"

File: 2025/12/03/structured-outputs-llms-reliable-json/typescript/src/schema.ts
import { z } from 'zod';

/** Task category enum */
export const CategorySchema = z.enum(['bug', 'feature', 'question', 'other']);
export type Category = z.infer<typeof CategorySchema>;

/** Task triage schema for categorizing and prioritizing issues */
export const TaskTriageSchema = z.object({
  category: CategorySchema,
  priority: z.number().int().min(1).max(5),
  needs_human: z.boolean(),
  summary: z.string().optional(),
}).strict(); // Reject extra fields

export type TaskTriage = z.infer<typeof TaskTriageSchema>;

/** User information extraction schema */
export const UserInfoSchema = z.object({
  name: z.string(),
  email: z.string().email(),
  age: z.number().int().min(0).max(150).optional(),
}).strict();

export type UserInfo = z.infer<typeof UserInfoSchema>;

